### ISSUE TITLE
Support for OFFSET Pushdown in Custom Trino Connector

### ISSUE THREAD

**Issue by lihonosov** [trinodb/trino#23963]
I’m working on a custom Trino connector (query external data source via HTTP protocol) and would like to push down the `OFFSET` parameter directly to the data source. However, it seems that Trino might not support this functionality fully at the moment.

For example, when I run a query like:
```SQL
SELECT * FROM custom_connector_table OFFSET 1 LIMIT 1
```

the [ConnectorMetadata.applyLimit](https://github.com/trinodb/trino/blob/master/core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java#L1299) method only receives `limit = 2`, without a clear way to distinguish between the `OFFSET` and `LIMIT` values separately.

Is there a way to directly access the values of both `OFFSET` and `LIMIT` in the current Trino API, or is there another recommended approach for implementing OFFSET **pushdown** in a custom connector?

Any guidance on this would be greatly appreciated!

**Comment by chenjian2664** [comment]
@hashhar @kokosing @ebyhr 
Do you think it is worth to add a rule to pushdown the `offset with limit`, if yes I wanna take this

**Comment by lihonosov** [comment]
It also happens with the [Postgres connector](https://trino.io/docs/current/connector/postgresql.html):

![Image](https://github.com/user-attachments/assets/d3a91612-cd03-4283-8754-676be6ccac14)

![Image](https://github.com/user-attachments/assets/8783dcca-78ac-4361-985c-9f5c829ea142)

Query log from Postgres:
![Image](https://github.com/user-attachments/assets/d2b36c02-e207-43ec-8883-2c93a306d610)


So, instead of returning **1** row, Postgres will return up to **1,001** rows

**Comment by chenjian2664** [comment]
It's because of the `PushLimitThroughOffset`.

**Comment by chenjian2664** [comment]
@lihonosov What kind of connector are you use?

**Comment by hashhar** [comment]
The concept of `OFFSET` being pushed down is hard because a connector might generate multiple splits and then the engine will not be able to reconstruct the combined results in the correct order.

This is the same reason why ORDER BY cannot be pushed down safely (other than assuming data for each split is already ordered).

**Comment by hashhar** [comment]
we can however try to enhance `PushLimitThroughOffset` to also pushdown the offset values in cases where a sort order is also present. That should be safe to do.

BTW are you trying to do pagination of results using OFFSET + LIMIT? If yes then there's better ways and the optimisation of pushdown is merely a bandaid.

**Comment by chenjian2664** [comment]
For what I have understand now, the jdbc are using the FixedPageSource, so it's safe to pushdown the OFFSET in some of the jdbc connectors, or I am missing something? @hashhar

**Comment by lihonosov** [comment]
> The concept of OFFSET being pushed down is hard because a connector might generate multiple splits and then the engine will not be able to reconstruct the combined results in the correct order.

yeah, but in our case, we have a very simple use case with a relatively small response, so the implementation always defines a single split

> The other problem is OFFSET without an ORDER BY is non-deterministic which makes the splits non-deterministic too which means FTE will need to be disabled for such connectors.

You might be querying an external REST service and that service might use classic parameters like `offset` and `limit` (or `page` and `size`), with the ordering handled within that service itself, ensuring the result is deterministic

> BTW are you trying to do pagination of results using OFFSET + LIMIT? If yes then there's better ways and the optimisation of pushdown is merely a bandaid.

In one example, yes, `OFFSET + LIMIT` is used for pagination. Since it's a small table, it isn't a problem. I completely agree that using classic pagination isn't a good choice for large or huge tables. In such cases, it's better to use something like a scrolling or cursor-based API if that's what you meant. 

Another example is pushing down a full SQL query to another SQL-like engine. For instance:
```sql
WITH users AS (
    SELECT id, name FROM external.sql.service OFFSET 100 LIMIT 100
)
SELECT u.id, u.name, a.value
FROM users AS u
LEFT JOIN catalog.abc.activity AS a ON (u.id = a.id)
```
This is just a made-up example, but what's the point of loading 200 records if Trino will drop the first 100 anyway?

It feels that solving this issue would cost more than the benefits it might bring. However, I just wanted to highlight it to better understand how Trino works 😃

**Comment by hashhar** [comment]
Yes, the main trouble is that the optimisation has to make sense for some connector that exists in Trino, not a custom out-of-tree connector for it to be worth the effort of adding and maintaining it over time.

For JDBC connectors this seems useful at first glance but the queries with OFFSET can be rewritten to better forms in that case.

For other connectors (e.g. APIs as you pointed out) this rewrite might not be possible.

If you use your connector to abstract an API have you considered what trino-rest does by using prediacates (which can optionally be mandatory) to pass parameters like offset or page? That would make pushdown much easier too and supported for more query shapes (e.g. JOINs).
